src/lib/crypto.cpp

```diff
@@ -102,6 +102,20 @@ pgp_generate_seckey(const rnp_keygen_crypto_params_t *crypto,
     seckey->tag = primary ? PGP_PTAG_CT_SECRET_KEY : PGP_PTAG_CT_SECRET_SUBKEY;
     rng = crypto->rng;
 
+    if (seckey->alg == PGP_PKA_ECDH) {
+        if (!ecdh_set_params(&seckey->material.ec, crypto->ecc.curve)) {
+            RNP_LOG("Unsupported curve [ID=%d]", crypto->ecc.curve);
+            goto end;
+        }
+        if (crypto->ecc.curve == PGP_CURVE_25519) {
+            if (x25519_generate(rng, &seckey->material.ec)) {
+                RNP_LOG("failed to generate x25519 key");
+                goto end;
+            }
+            seckey->material.ec.curve = crypto->ecc.curve;
+            goto after_seckey_alg;
+        }
+    }
```

Comment:
No need to move this outside from `switch()`. This makes code less readable and more complicated.
It is enough to add /* FALLTHROUGH */ or similar comment, please see the details here: https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wimplicit-fallthrough